torch/nn/modules/module.py

```diff
@@ -382,6 +382,141 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
         self._modules[name] = module
 
+    def has_submodule(self, target: str) -> bool:
+        """
+        Returns whether or not this Module contains the submodule
+        given by ``target``.
+
+        For example, let's say you have an ``nn.Module`` ``A`` that
+        looks like this:
+
+        .. code-block::text
+
+            A(
+                (net_b): Module(
+                    (net_c): Module(
+                        (conv): Conv2d(16, 33, kernel_size=(3, 3), stride=(2, 2))
+                    )
+                    (linear): Linear(in_features=100, out_features=200, bias=True)
+                )
+            )
+
+        (The diagram shows an ``nn.Module`` ``A``. ``A`` has a nested 
+        submodule ``net_b``, which itself has two submodules ``net_c`` 
+        and ``linear``. ``net_c`` then has a submodule ``conv``.)
+
+        To check whether or not we have the ``linear`` submodule, we
+        would call ``has_submodule("net_b.linear")``. To check whether
+        we have the ``conv`` submodule, we would call
+        ``has_submodule("net_b.net_c.conv")``.
+
+        Args:
+            target: The fully-qualified string name of the submodule
+                to look for. (See above example for how to specify a
+                fully-qualified string.)
+
+        Returns:
+            bool: Whether or not the target string referenced an
+                existing submodule. Returns False if the target string
+                resolves to something that is not an ``nn.Module``.
+        """
+        atoms: List[str] = target.split(".")
+
+        for item in atoms:
+
+            if not hasattr(self, item):
+                return False
+
+            self = getattr(self, item)
+
+            if not isinstance(self, torch.nn.Module):
+                return False
+
+        return True
+
+    def add_submodule(self, target: str, m: 'Module') -> bool:
+        """
+        Adds the given submodule to ``self``.
+
+        This installs empty Modules where none exist yet if they are 
+        subpaths of ``target``.
+
+        Args:
+            target: The fully-qualified string name of the new submodule
+                (See example in ``has_submodule`` for how to specify a
+                fully-qualified string.)
+            m: The submodule itself; the actual object we want to
+                install in the current Module
+
+        Return:
+            bool: Whether or not the submodule could be inserted. For
+                this method to return True, each object in the chain
+                denoted by ``target`` must either a) not exist yet,
+                or b) reference an ``nn.Module`` (not a parameter or
+                other attribute)
+
+        """
+        *prefix, field = target.split('.')
+        mod: torch.nn.Module = self
+
+        for item in prefix:
+
+            submod = getattr(mod, item, None)
+
+            if submod is None:
+                submod = torch.nn.Module()
```

Comment:
Adding a base `Module` here has some problems. It is generally intended that modules derive from `nn.Module` and implement the `forward()` function for their computation. For the base `Module` added here, there will be no `forward()` implemented, so the submodule won't be usable without some hackery.

```python
import torch
from torch import nn

class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.add_submodule('sub.linear1', nn.Linear(4, 3))
        self.add_submodule('sub.linear2', nn.Linear(3, 2))
        self.add_submodule('sub.linear3', nn.Linear(2, 1))

    def forward(self, x):
        return self.sub(x)

m = MyModule()
print(m(torch.randn(10, 4)))  # throws NotImplementedError
```